src/v/raft/configuration_manager.cc

```diff
@@ -433,6 +433,39 @@ model::revision_id configuration_manager::get_latest_revision() const {
     return _configurations.rbegin()->second.cfg.revision_id();
 }
 
+int64_t configuration_manager::offset_delta(model::offset o) const {
+    auto it = lower_bound(o);
+
+    if (it == begin()) {
+        /**
+         * iterator points to the first configuration with offset greater
+         * than then requsted one. Knowing an index of that configuration we
+         * know that there was exactly (index -1) configurations with offset
+         * lower than the current one. We can simply subtract one from
+         * index.
+         */
+        return std::max<int64_t>(0, it->second.idx() - 1);
+    }
+
+    return std::prev(it)->second.idx();
+}
+
+ss::future<> configuration_manager::adjust_configuration_idx(
+  configuration_idx new_initial_idx) {
+    return _storage.kvs()
+      .put(
+        storage::kvstore::key_space::consensus,
+        next_configuration_idx_key(),
+        reflection::to_iobuf(new_initial_idx))
+      .then([this, new_initial_idx] {
+          auto idx = new_initial_idx;
+          for (auto& [_, cfg] : _configurations) {
+              cfg.idx = idx++;
```

Comment:
As I understand it this is resetting all configurations' offsets back to offsets at the start of the log, which presumably makes the resulting state valid but doesn't actually reflect the offsets where the configurations really appear?

This seems reasonable as a recovery mechanism for corrupted systems but I'm not deep enough into this code to be sure that it couldn't have side effects.  No objection to merging though.

Reply:
This is exactly what it does, information about actual offset is lost, the best what we can do is to adjust all of the offsets to `0`. This code will only trigger when log needs an adjustment. It will not influence other cases.